fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/CustomItemSetting.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.item.v1;
+
+import java.util.function.Supplier;
+
+import org.jetbrains.annotations.ApiStatus;
+
+import net.minecraft.item.Item;
+
+import net.fabricmc.fabric.impl.item.CustomItemSettingImpl;
+
+/**
+ * A type of setting that can be passed to item constructors.
+ * This feature can be used by mods to add non-standard settings
+ * to items in a way that is compatible with other mods that add
+ * settings to items.
+ *
+ * <p>Values of this setting can be retrieved from an item using {@link CustomItemSetting#getValue(Item)}</p>
```

Comment:
I would suggest adding, in the javadoc:
- Something along the lines of "Values are directly stored inside this `CustomItemSetting`, per instance. Users that wish to expose a custom setting may do so by exposing the `CustomItemSetting` instance, for example in a `public static final` field."
- Maybe a little example wrapped in `<pre>{@code ... }</pre>` to follow that up and make usage a bit more obvious, a `CustomItemSetting<String>` like the testmod would do I think.